### PR TITLE
Add firewall rules

### LINKED ISSUE #1723: Server installer - add firewall rules
## Is your feature request related to a problem? Please describe 👀

When running the server on a windows machine - trying to connect to the server from another computer required the unblocking of the server port (`8000` by default). It would be good if the installer was able to do this.

When the desktop version is installed and run for the first time, it asks you to accept some changes to the windows firewall. It would be better if the installer made those changes.

## Describe the solution you'd like 🎁
For the desktop and standalone installer:
- [ ] Create the rule to allow the application

For the server and standalone installer:
- [ ] Allow the server port

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #1723 

# 👩🏻‍💻 What does this PR do? 
 
Adds an incoming firewall rule for the server port to both the server and standalone installers.
For the rule to be applied, the firewall service requires restarting, so that happens too.
On uninstall, the rules are removed.

For the standalone and desktop installers I've added the rules which were added when the app was first run - I thought that if we add the rules, then the user won't get prompted.

# 🧪 How has/should this change been tested? 

Have created installers and tested on a couple of different machines. 

## 📃 Documentation

_No user facing changes_

### CODE REVIEW

File: build/windows/omsupply_server.suf [hunk 1]
```diff
@@ -4276,6 +4279,21 @@ end
 
 File.Delete(sFileToBeDeletePath); -- Delete unchoosen app file (.exe) by a user
 
+-- Add a firewall rule to allow incoming requests
+local firewallRuleKey = "SYSTEM\\CurrentControlSet\\Services\\SharedAccess\\Parameters\\FirewallPolicy\\FirewallRules";
+if (Registry.DoesKeyExist(HKEY_LOCAL_MACHINE, firewallRuleKey)) then
+
+	Registry.SetValue(HKEY_LOCAL_MACHINE, firewallRuleKey, "open-mSupply-Server-TCP-IN", "v2.28|Action=Allow|Active=TRUE|Dir=In|Protocol=6|LPort=" .. SessionVar.Get("%omsupply_port%") .. "|Name=omSupply Server|", REG_SZ);
```
Comment (andreievg) [inline build/windows/omsupply_server.suf:4286]: Hmm shouldn't Name also come from SessionVar as ProductName

**Review by andreievg — COMMENTED** [review]
Were this scripts auto generated by setup factory or hand crafted ? 

I agree that we should create rule for incoming omsupply_port for server. But I like the message that's given when desktop client app starts to run, I think it makes the user feel in control (safer)

I had a quick look at the script, if it's hand crafted can you please point me to the documentation (or even better add as comment to the script). I can't quite see the restriction by subnet, is it possible to restrict by local network ?

**Comment by mark-prins** [comment]
> Were this scripts auto generated by setup factory or hand crafted ?

Setup factory - I updated them on jenkins and built a version on there to test.
I'll remove the desktop rules; it's fine to have them added explicitly by the user.

**Comment by mark-prins** [comment]
> But I like the message that's given when desktop client app starts to run, I think it makes the user feel in control (safer)

Strangely, I'm the opposite. When windows pops up with an alert to say that something about this app has been blocked! do you want to allow it? that makes me feel less safe, like windows has prevented something potentially unsafe from running.

File: build/windows/omsupply_server.suf (hunk 1 above)
Comment (mark-prins) [reply build/windows/omsupply_server.suf]: updated with f13595469

**Comment by mark-prins** [comment]
@andreievg ok, you win. Have removed the desktop rules in 03ca6bbc7

The server rule worked fine, but the desktop still prompted even with those rules in place. So I've taken them out.

File: build/windows/omsupply_server.suf (hunk 1 above)
Comment (andreievg) [reply build/windows/omsupply_server.suf]: > it's just a firewall rule so I didn't bring that through

Ahh ok, i saw that it's been done for port `.. SessionVar.Get("%omsupply_port%") ..`

> updated with https://github.com/openmsupply/open-msupply/commit/f135954691e6d179b6ad1b3aa797c83d07fcf4f3

Thanks

**Comment by andreievg** [comment]
> Strangely, I'm the opposite. When windows pops up with an alert to say that something about this app has been blocked! do you want to allow it? that makes me feel less safe, like windows has prevented something potentially unsafe from running.

Hmm, i asked Nick if it was an issue in his support experience, happy to revert if needed

**Review by andreievg — CHANGES_REQUESTED** [review]
Thank you.

> Strangely, I'm the opposite. When windows pops up with an alert to say that something about this app has been blocked! do you want to allow it? that makes me feel less safe, like windows has prevented something potentially unsafe from running.

Nick said it would reduce the chance of possible issues (if Desktop firewall rule is embedded in installer). My worry was it's hand crafter script that may be hard to validate/test, but since it's provided by setup factory, happy to have it reinstated.

One last change thought, `Profile=Private`:

https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-gpfas/2efe0b76-7b4a-41ff-9050-1023f8196d16

```
Profile=: This token value represents the dwProfiles field of the FW_RULE structure. The PROFILE-VAL grammar rule represents a value content of such field. If this token appears more than once in a RULE grammar rule, then all the contents represented by the PROFILE-VAL rule appearing next to them are included. If the Profile= token never appears in the rule string then it represents a value of FW_PROFILE_TYPE_ALL as defined in [MS-FASP] section [2.2.2](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-fasp/7704e238-174d-4a5e-b809-5f3787dd8acc).
```

**Comment by mark-prins** [comment]
I've tested and the firewall prompt is something that we cannot control by adding rules.
Windows adds a rule with `Action=Block` on install and the prompt is to change that to `Allow`
Best, I think, to merge this and get the server port opened. We can investigate the desktop firewall prompt later if that's an issue.

**Review by andreievg — APPROVED** [review]
> I've tested and the firewall prompt is something that we cannot control by adding rules.
Windows adds a rule with Action=Block on install and the prompt is to change that to Allow
Best, I think, to merge this and get the server port opened. We can investigate the desktop firewall prompt later if that's an issue.

Ok, glad that we checked, good to not have extra, unused config in installer.

Good to merge although the change request was for `Profile=Private` on server firewall setting (technically might be ok without it, since most routers would already restrict this)

**Comment by mark-prins** [comment]
oh ok. I was unsure what you meant by the Profile=Private comment. I'll have to rebuild and retest to check that change.

**Pull request merged**